web/source/dom/targets/input.ts

```diff
@@ -54,53 +58,58 @@ namespace com.keyman.dom.targets {
     }
 
     getCaret(): number {
-      if(this.root.selectionStart == this._cachedSelectionStart) {
-        return this.processedSelectionStart;
-      } else {
+      if(this.root.selectionStart != this._cachedSelectionStart) {
         this._cachedSelectionStart = this.root.selectionStart; // KMW-1
         this.processedSelectionStart = this.root.value._kmwCodeUnitToCodePoint(this.root.selectionStart); // I3319
         this.processedSelectionEnd = this.root.value._kmwCodeUnitToCodePoint(this.root.selectionEnd);     // I3319
-
-        return this.processedSelectionStart;
       }
+      return this.root.selectionDirection == 'forward' ? this.processedSelectionEnd : this.processedSelectionStart;
     }
 
     getDeadkeyCaret(): number {
       return this.getCaret();
     }
 
     setCaret(caret: number) {
-      this.setSelection(caret, caret);
+      this.setSelection(caret, caret, "none");
     }
 
-    setSelection(start: number, end: number) {
+    setSelection(start: number, end: number, direction: "forward" | "backward" | "none") {
       let domStart = this.root.value._kmwCodePointToCodeUnit(start);
       let domEnd = this.root.value._kmwCodePointToCodeUnit(end);
-      this.root.setSelectionRange(domStart, domEnd);
+      this.root.setSelectionRange(domStart, domEnd, direction);
 
       this.processedSelectionStart = start;
       this.processedSelectionEnd = end;
 
-      keyman.dom.Utils.forceScroll(this.root);
+      Utils.forceScroll(this.root);
+    }
+
+    getSelectionDirection(): "forward" | "backward" | "none" {
+      return this.root.selectionDirection;
     }
 
     getTextBeforeCaret(): string {
-      return this.getText()._kmwSubstring(0, this.getCaret());
+      this.getCaret();
+      return this.getText()._kmwSubstring(0, this.processedSelectionStart);
     }
 
     setTextBeforeCaret(text: string) {
       this.getCaret();
+      let selectionLength = this.processedSelectionEnd - this.processedSelectionStart;
```

Comment:
Does this have to take `selectionDirection` into consideration so that we don't end up with a negative length?

Reply:
No -- see my previous comment.